feat(events): consolidate to events pkg, remove unnecessary code

rvagg opened the pull request:
Closes: https://github.com/filecoin-project/lassie/issues/17

I _think_ #17 is done with this. It was mostly done in #25 but this does some additional cleanup. Here's the state of things:

* The client no longer emits events of its own, it'll just allow you to hook into datatransfer events
* datatransfer events stop at retrieve.go, where they are used, along with other information, to construct native lassie events from the new events package.
* The `RetrieveFromCandidates()` call in retrieve.go, where most of the gnarly magic happens to fetch a CID, receives an optional `func(events.RetrievalEvent)` callback which gets a stream of these events. Those events all come in on the main goroutine that runs the retrieval (which sits in `collectResults` until done, from there it dispatches events as well as ticks down all of the candidates' goroutines as they complete).
* The `Retriever`, the main entrypoint we expect users to use to manage the lifecycle of multiple retrievals has an `events.EventManager`, which is an asynchronous thing which will suck in all of those `events.RetrievalEvent` and then push the out along with the unique `RetrievalID` (a UUID) to registered `events.RetrievalEventListener`s.
* `events.RetrievalEventListener` doesn't actually receive `events.RetrievalEvent` objects, it implements a function for each broad event category - mainly "progress", "failure" and "success" for each phase, (although the [new] indexer phase's "success" is actually "IndexerCandidates"). The "progress" calls have an event `Code` to distinguish the details.
* For our event recording, we end up picking them apart and pushing them, along with `Code` to the HTTP endpoint.

The piece I think might need attention is the `events.RetrievalEventListener` interface: https://github.com/filecoin-project/lassie/blob/rvagg/events-consolidation/pkg/events/events.go#L55
* We end up picking apart `events.RetrievalEvent` objects by sub-type: https://github.com/filecoin-project/lassie/blob/rvagg/events-consolidation/pkg/retriever/retriever.go#L255-L279 so that we can fire appropriate events on that interface to listeners
* You can see it being used fairly simply here: https://github.com/filecoin-project/lassie/blob/rvagg/events-consolidation/cmd/lassie/fetch.go#L194-L234
* BUT if you look in the eventrecorder that sends data off to our HTTP endpoint: https://github.com/filecoin-project/lassie/blob/rvagg/events-consolidation/pkg/eventrecorder/eventrecorder.go - and compare that to what actually gets recorded in our database; we mostly end up unwinding the categorisation of these events into these broad "progress", "success" and "failure" calls and instead using the `event.Code` to record, which ends up looking quite a lot like the `events.RetrievalEvent` objects.

SO another option with all of this is to ditch that intermediate interface and just allow a user to subscribe to a simple stream of raw `events.RetrievalEvent` objects and force them to do a big 'ol `switch event.(type)` on them to pick out the details. I did originally implement that way back when I did the eventrecorder but decided it was ugly and messy and would be cleaner to implement a multi-method listener interface. But perhaps that was the wrong call.

I know that's a lot of information. If I feel inspired I might go and implement the second option and propose that to see what it looks like.

rvagg commented:
Latest commit takes this a step further and removes the dual eventing layering. Previously there were streams of retrieval events coming in to the Retriever which would then push them out via the EventManager to the the EventListeners, differentiating the event types into separate calls on those listeners. Now, the EventListener interface is replaced with an EventSubscriber function which just receives a stream of events. Those events can be differentiated either by `Code()` or by type (some of the types have extra methods to get more info). You can see the result of this in the lassie fetch command code, specifically here: https://github.com/filecoin-project/lassie/pull/41/files#diff-176ffde30ef6db7af6cca7002660ab88290d9fef85c236ed075e3ea439fceef1R197 - a big `switch` that does something different for the event types.

There is still some goroutine juggling - the events come in from the various retrieval event goroutines into the goroutine that originally called the retrieval (`collectResults` does this), and then they get further consolidated into the main `Retriever` and pushed to the single EventManager which does dispatch on a dedicated goroutine. So you're guaranteed to get events on a single goroutine, but it's not the one you're retrieving from. This was originally done when we did the event publisher (which has further goroutines going on to manage things!) and it's something we could revisit. Perhaps having the events come in on the same goroutine where you called `Retrieve()` would be best? We could also consider a model similar to how we've done go-data-transfer now where you just provide a callback rather than a global listener, or both. Because now one `Retriever` will give subscribers events across all retrievals, differentiated by RetrievalID.

kylehuntsman commented on pkg/events/manager.go:
```diff
@@ -0,0 +1,114 @@
+		ctx:         ctx,
+		cancel:      cancel,
+		subscribers: make(map[int]types.RetrievalEventSubscriber),
+		events:      make(chan types.RetrievalEvent, 16),
```
Was there rationale around 16 as the buffer size here? Just something big enough to not block?

hannahhoward replied:
That is my guess. In the context where you're handing off to arbitrary user defined subscribers that could be time consuming, even this my not be enough to gaurantee not blocking.

@rvagg have you looked at the code that ultimately powers the subscriber interfaces on data transfer -- it might be worth borrowing from -- we originally had things setup with just a buffer but found cases where we could exceed the buffer.

https://github.com/filecoin-project/go-statemachine/blob/master/fsm/fsm.go#L159

Nowadays I'd probably use a linked list for this since I worry golang slices would grow arbitrarily here.

hannahhoward replied:
ultimately this may be premature observation but just wanted to make you all aware of the concern

kylehuntsman replied:
No no, I had the same concern. My thought was if this is not enough, we'd need a data structure to dump the events into. A linked list wouldn't be awful.

kylehuntsman reviewed (approved):
I do like the switch case interface more than the previous iterations. Moving the event types to the types package is a nice change. I left a couple questions, otherwise I like it.

hannahhoward commented on pkg/events/events.go:
```diff
@@ -0,0 +1,314 @@
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var (
+	_ types.RetrievalEvent = RetrievalEventCandidatesFound{}
+	_ types.RetrievalEvent = RetrievalEventCandidatesFiltered{}
+	_ types.RetrievalEvent = RetrievalEventConnected{}
+	_ types.RetrievalEvent = RetrievalEventQueryAsked{}
+	_ types.RetrievalEvent = RetrievalEventQueryAskedFiltered{}
+	_ types.RetrievalEvent = RetrievalEventProposed{}
+	_ types.RetrievalEvent = RetrievalEventAccepted{}
+	_ types.RetrievalEvent = RetrievalEventFirstByte{}
+	_ types.RetrievalEvent = RetrievalEventFailed{}
+	_ types.RetrievalEvent = RetrievalEventSuccess{}
+)
```
@kylehuntsman

This type of structure is used to verify that the concrete type in question (i.e. `MyThing{}` in your example) satisfies the interface in question (`type` in your example). If any of the interface methods are missing or incorrect on your concrete type, this statement will generate a compile error.

It's basically a programmer trick to have the compiler tell you if you forgot to implement a method or something before you try to use the concrete type somewhere else, kind of a type test inside your code to make sure you did it right.

hannahhoward replied:
I will say, as someone who's programmed for a minute, I basically try to constantly have the compiler tell me if I'm doing it wrong, cause I am almost always doing it wrong :P

hannahhoward commented on pkg/eventrecorder/eventrecorder.go:
```diff
@@ -86,61 +82,40 @@ type EventDetailsError struct {
+	switch ret := event.(type) {
+	case events.EventWithCandidates: // events.RetrievalEventCandidatesFound, events.RetrievalEventCandidatesFiltered:
+	case events.RetrievalEventConnected:
+	case events.EventWithQueryResponse: // events.RetrievalEventQueryAsked, events.RetrievalEventQueryAskedFiltered:
```
huh! I didn't know you could switch an interface to an interface but I guess you can.

rvagg replied:
yeah, kind of weird, but works: https://go.dev/play/p/2LbRnZs0BU
this branch also shows up in coverage, so it's being exercised

hannahhoward commented on pkg/events/manager.go:
```diff
@@ -0,0 +1,114 @@
+// dispatched. A channel is returned that will receive a single value when the
+// event loop has started.
+func (em *EventManager) Start() chan struct{} {
+	startChan := make(chan struct{})
+	go func() {
+		em.lk.Lock()
+		em.started = true
+		em.lk.Unlock()
+		startChan <- struct{}{}
```
@rvagg I agree with @kylehuntsman . I'm not sure what startChan is therefore to begin with. I think maybe it's cause you're setting started=true inside the go routine? But I think you should just start the go routine and then set started=true in regular go-routine that called the function.

i.e:

```golang
func (em *EventManager) Start() {
   go func() {
      for {
          // ...
      }
   }()
   em.lk.Lock()
   em.started=true
   em.lk.Unlock()
}
```

Can you explain what I'm missing?

hannahhoward commented on pkg/retriever/retriever_test.go:
```diff
@@ -301,9 +357,9 @@ func TestRetriever(t *testing.T) {
+			ret.RegisterSubscriber(subscriber.Collect)
 
 			// --- start ---
 			select {
```
again, is there a concrete case where the retriever doesn't start? It just appears to be calling eventmanager.Start() which doesn't appear to have a case where it doesn't start really.

hannahhoward reviewed (commented):
I like these changes overall.

I do want to  either eliminate the complexity around returning a channel from eventManager.Start() before we merge, or understand why it is neccessary.

rvagg commented:
Feedback addressed, PTAL.

1. The `Start()` thing I don't exactly remember the original reason for it but IIRC it was breaking CI because of the quick tests around events - the start/stop cycle being too quick to capture anything, or something like that, probably Windows. It also may have been a desire to have symmetric start and stop interfaces. :shrug: anyway I've removed it now and CI is currently passing so that's good I suppose.
2. The event queue thing, including the channel with size of `16` isn't new code, it's just been moved and adjusted. It comes from the original event work in https://github.com/application-research/autoretrieve/pull/97 and the `16` specifically comes from feedback on that PR, one of the last commits in it that was addressing concerns about event backlog: <https://github.com/application-research/autoretrieve/pull/97/commits/f44fbe3c13ad42132cf24e7996bdb3f83e69dc4a>. But anyway, I've gone ahead and implemented a non-channel queue system like in fsm, but with a linked-list. Please check it out to sanity check my work! It's neat, but also probably overkill because the event reporter has its own batching system.
